artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java

```diff
@@ -467,6 +478,83 @@ private boolean internalAddressInfo(AddressInfo addressInfo, boolean reload) thr
+         if (internalQueue != null && internalQueue.getRingSize() != settings.getRetroactiveMessageCount()) {
+            internalQueue.setRingSize(settings.getRetroactiveMessageCount());
+         }
+      };
+      addressSettingsRepository.registerListener(repositoryChangeListener);
+      server.getAddressInfo(name).setRepositoryChangeListener(repositoryChangeListener);
+   }
+
+   private void createRetroactiveResources(final SimpleString retroactiveAddressName, final long retroactiveMessageCount, final boolean reload) throws Exception {
+      String prefix = server.getConfiguration().getInternalNamingPrefix();
+      final SimpleString internalAddressName = ResourceNames.getRetroactiveResourceName(prefix, retroactiveAddressName, ResourceNames.ADDRESS);
+      final SimpleString internalQueueName = ResourceNames.getRetroactiveResourceName(prefix, retroactiveAddressName, ResourceNames.QUEUE);
+      final SimpleString internalDivertName = ResourceNames.getRetroactiveResourceName(prefix, retroactiveAddressName, ResourceNames.DIVERT);
+
+      if (!reload) {
+         AddressInfo addressInfo = new AddressInfo(internalAddressName).addRoutingType(RoutingType.ANYCAST).setInternal(true);
+         addAddressInfo(addressInfo);
+         server.createQueue(internalAddressName,
+                            RoutingType.ANYCAST,
+                            internalQueueName,
+                            null,
+                            null,
+                            true,
+                            false,
+                            false,
+                            false,
+                            false,
+                            0,
+                            false,
+                            false,
+                            false,
+                            0,
+                            null,
+                            false,
+                            null,
+                            false,
+                            0,
+                            0L,
+                            false,
+                            0L,
+                            0L,
+                            false,
+                            retroactiveMessageCount);
+      }
+      server.deployDivert(new DivertConfiguration()
+                             .setName(internalDivertName.toString())
+                             .setAddress(retroactiveAddressName.toString())
+                             .setExclusive(false)
+                             .setForwardingAddress(internalAddressName.toString())
+                             .setRoutingType(ComponentConfigurationRoutingType.STRIP));
+   }
+
+   private void removeRetroactiveResources(SimpleString address) throws Exception {
+      String prefix = server.getConfiguration().getInternalNamingPrefix();
+
+      SimpleString internalDivertName = ResourceNames.getRetroactiveResourceName(prefix, address, ResourceNames.DIVERT);
```

Comment:
Why divert and separate address is needed. Why not have the retro queue under the original address?
With having separate address will mean message has to be copied rather than it simply being a message reference to existing message.

Reply:
The more i think on this the more strongly i think we shouldnt have the divert and separate address, simply should be queue under the existing address.